Add non-halting exception support to resources

jerryaldrichiii opened the pull request:
This adds two `Inspec::Exceptions` that can be used within resources to either skip or fail a test without halting execution.

This fixes #1205.

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Add non-halting exception support to resources

This adds two `Inspec::Exceptions` that can be used within resources to
either skip or fail a test without halting execution.

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

adamleff commented on lib/inspec/plugins/resource.rb:
```diff
@@ -65,6 +71,14 @@ def skip_resource(message)
         end
 
+        def resource_failed
+          @resource_failed if defined?(@resource_failed)
```
If `@resource_failed` is never assigned a value, it will return `nil` by default. Do we need this `defined?` check?

jerryaldrichiii replied:
I agree with you, I honestly just repeated what was there. I started down the path of correcting it and noticed that quite a few unrelated tests checked the value of `resource_skipped` and would need to be changed to `resource_skipped_message`.

The real worry here is that some of these tests are Unit tests for resources. Which got me thinking, if I modify `resource_skipped` to be a `?` and `_message` would that be a breaking change for some?

If so, should I:
  - Fix `resource_failed` only and have it be different than `resource_skipped` (and put in an issue for InSpec 2.0 to correct the difference)
  - Just leave it as is and put in an issue for InSpec 2.0 to fix both?

Thoughts?

adamleff replied:
Just because something was there doesn't mean it's the right thing to do. :) We should always look for improvements while mucking around in the codebase.

The question to ask is: "are these methods part of an expected public API?" Is `resource_failed` used anywhere other than this class and its unit tests? Do the consumers of this method extend beyond the InSpec codebase (i.e. can a user call these methods from within their tests, and is it something we support)? And, if the answer is yes to either of these, can we do something to transition to what we want it to be?

If you agree that having `?` methods is a better fit here, then I'd like to encourage you to think about ways to get there without introducing breaking changes.  For example:

 - add the `?` methods
 - add the new `_message` methods
 - create method aliases for the non-`?` methods to the `_message` methods

I think you'll find it you do that, you gain a more idiomatic implementation while maintaining backward compatibilities (and we should add code comments explaining why the method aliases exist) and then we can remove those aliases in the future.

jerryaldrichiii replied:
Excellent suggestions! I didn't think about an alias method. I pushed up another commit implementing what you recommended.

It also modifies other files that used `resource_skipped` in order to reduce work in the future when we deprecate. Let me know if I need to remove those changes.

adamleff reviewed (changes requested):
This is a great improvement. Small nitpick re: an unnecessary `defined?` check.

Also, I wonder if it would be cleaner to have something like this:

```ruby
def resource_failed?
  !@resource_failed.nil?
end

def resource_failure_message
  @resource_failed
end

def resource_skipped?
   # ... etc ...
end
```

It feels a bit better to me to have a boolean method for checking and only grab the message when necessary. This isn't a blocker though -- just wanted to open a conversation about it with you.

adamleff commented on lib/inspec/exceptions.rb:
```diff
@@ -5,6 +5,8 @@ module Inspec
     class AttributesFileNotReadable < ArgumentError; end
+    class ResourceFailed < StandardError; end
     class SecretsBackendNotFound < ArgumentError; end
+    class ResourceSkipped < StandardError; end
```
Can we keep this list alphabetized for consistency?

jerryaldrichiii replied:
Hah! I originally had them called something else and took great care to make sure they were alphabetical. I renamed to both be `Resource` plus result so they would be together and forgot to sort.

Great catch @adamleff! Fixed in latest commit.

jerryaldrichiii committed:
Modify exception order to be alphabetical

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Add `?` and `_message` to resource skipped/failed

This removes `resource_skipped` but give the same functionality via an
alias method. This also replaces its usage throughout the codebase to
reduce work when deprecating in the future.

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

adamleff commented on lib/inspec/plugins/resource.rb:
```diff
@@ -57,14 +67,34 @@ def self.example(example = nil)
+        end
+        # This alias exists for backwards compatibility
+        # It is more idiomatic to have a `?` and `_message` methods
+        # TODO: Remove in InSpec 2.0
```
If you want to remove in 2.0, we need to change this to a deprecation warning now. :)

```
def resource_skipped
  warn '[DEPRECATED] Use `resource_skipped_message` for the resource skipped message. This method will be removed in InSpec 2.0.'
  @resource_skipped
end
```

... and then run through some manual tests to uncover if we are (or RSpec is) using it anywhere internally.

jerryaldrichiii replied:
Will do :smile:

adamleff commented on test/unit/profiles/profile_context_test.rb:
```diff
@@ -141,13 +141,13 @@ def load(call)
       it 'alters controls when positive' do
         profile.load(if_false + control)
         get_checks.length.must_equal 1
-        get_checks[0][1][0].resource_skipped.must_equal 'Skipped control due to only_if condition.'
```
Can I suggest a unit test that covers `resource_skipped` still, since we're leaving it in but deprecating it? That will help us in the future when we remove it, too.

jerryaldrichiii replied:
I added a unit test to `profile_resource_exceptions_test` to cover it. :+1:

jerryaldrichiii replied:
https://github.com/chef/inspec/pull/2235/commits/8f9d9b1240bc97080af614bd71ea6e4537d55368#diff-d175555c026cd3a023e95ac973639b31R65

adamleff commented on test/unit/profiles/profile_context_test.rb:
```diff
@@ -141,13 +141,13 @@ def load(call)
         profile.load(if_false + describe)
         get_checks.length.must_equal 1
-        get_checks[0][1][0].resource_skipped.must_equal 'Skipped control due to only_if condition.'
+        get_checks[0][1][0].resource_skipped_message.must_equal 'Skipped control due to only_if condition.'
```
Can we include some `resource_skipped?` and `resource_failed?` tests in these, too? It appears we're missing coverage on those.

jerryaldrichiii replied:
Can do! I'll add test cases for each and their messages.

adamleff commented on test/unit/profiles/profile_resource_exceptions_test.rb:
```diff
@@ -0,0 +1,64 @@
+  describe 'within initialize' do
+    it 'skips resource when `Inspec::Exceptions::ResourceSkipped` is raised' do
+      checks[0][0][1][0].resource_skipped_message.must_equal 'Skipping because reasons'
+      checks[0][0][1][0].resource_failed?.must_equal false
```
Should we be checking `resource_skipped?` here too?

jerryaldrichiii replied:
Yup. :smile: I'll add test cases for each and their messages.

adamleff commented on test/unit/profiles/profile_resource_exceptions_test.rb:
```diff
@@ -0,0 +1,64 @@
+
+    it 'fails resource when `Inspec::Exceptions::ResourceFailed` is raised' do
+      checks[1][0][1][0].resource_failed_message.must_equal 'Failing because reasons'
+      checks[1][0][1][0].resource_skipped?.must_equal false
```
Should we be checking `resource_failed?` here too?

jerryaldrichiii replied:
Yup. :smile: I'll add test cases for each and their messages.

jerryaldrichiii committed:
Disable `Metrics/MethodLength` for `__register`

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

adamleff reviewed (changes requested):
This is a huge improvement, @jerryaldrichiii. I have some questions/suggestions. Reach out if anything is unclear.

jerryaldrichiii committed:
Remove alias in favor of deprecation warning

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Add more resource exception tests

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

adamleff reviewed (approved):
@jerryaldrichiii this is great, thanks!

Before we merge, can you delete unneeded file `test/unit/mock/profiles/profile-with-resource-exceptions/libraries/.gitkeep` from this PR? Thanks!

jerryaldrichiii committed:
Remove unneeded `.gitkeep` file

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

chris-rock commented on lib/inspec/plugins/resource.rb:
```diff
@@ -57,12 +67,33 @@ def self.example(example = nil)
+          !@resource_skipped.nil?
+        end
+
+        def resource_skipped_message
```
Why do we need to distinguish between failed and skip message. Can we have both at the same time?

jerryaldrichiii replied:
Hmmm...you have a point. We shouldn't have both. Should we merge both into `resource_skipped_message`? Something else? Technically a failed resource isn't skipped since it tries to execute, but I don't want to rename it to `resource_message`. Thoughts?

adamleff replied:
Part of the approach here was to ensure that we weren't botching something that was considered to be "public" or something RSpec was expecting. I'm +1 for consolidating as long as it doesn't break any backward compatibility.

I think calling it `exception_message` is probably sufficient.

jerryaldrichiii replied:
I think merging into `exception_message` is a fine idea. Before, it was all contained in `resource_skipped` which should still work going forward.

jerryaldrichiii replied:
Well fiddlesticks...
https://github.com/chef/inspec/blob/master/test/unit/resources/nginx_conf_test.rb#L100

This tests calls the instance variable `@resource_skipped` expecting a message. Since this test does it, is it safe to assume that any user could call the instance variable directly instead of the deprecated `resource_skipped` method?

If so it's no big, I'll just keep `@resource_skipped` and `@resource_failed` set to messages and evaluate in `resource_exception_message`...but I'd really like those to be bools.

adamleff replied:
If you call an instance variable directly, you deserve breakage. 🙂  In no way can we support that as long-term behavior.

I'd fix the test and proceed with consolidation.

jerryaldrichiii replied:
Agreed! Fixed.

chris-rock commented on lib/inspec/plugins/resource.rb:
```diff
@@ -37,14 +37,24 @@ def __resource_registry
+            skip_resource(e.message)
+          rescue Inspec::Exceptions::ResourceFailed => e
+            fail_resource(e.message)
+          end
```
What happens in cases where we have an unhanded exception. I think we should handle those cases as well.

adamleff replied:
Ugh, can't believe I missed this. An `else raise` is totally appropriate here.

adamleff replied:
I take it back... Ruby will raise an exception if it's not one of the rescued classes. `else raise` is implicit. Given that we would just bubble up an exception anyways without this change (i.e. syntax issues in the initialize method), I think this is fine as-is.

adamleff replied:
In other words, we currently unwind InSpec if there's a ruby syntax issue inside a resource. We should continue to do that in my opinion.

chris-rock commented on lib/inspec/plugins/resource.rb:
```diff
@@ -37,14 +37,24 @@ def __resource_registry
-          super(*args)
+          begin
+            super(*args)
+          rescue Inspec::Exceptions::ResourceSkipped => e
```
It is great to see the separation here! Well done!

chris-rock commented on lib/inspec/plugins/resource.rb:
```diff
@@ -37,14 +37,24 @@ def __resource_registry
+    def __register(name, obj) # rubocop:disable Metrics/MethodLength
       cl = Class.new(obj) do
         def initialize(backend, name, *args)
+          @resource_skipped = nil
```
Should we use an enum? Essentially we have stopped the execution and we did this for different reasons. The reason is important to say if this is okay or not.

chris-rock reviewed (commented):
@jerryaldrichiii This is a huge improvement and it is very beneficial. I just added a comment around msg handling and the how we can support adding more reasons easily in the future without breaking the api.

jerryaldrichiii committed:
Merge skipped/failed messages and tweak predicates

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

chris-rock reviewed (approved):
Thank you for this improvement @jerryaldrichiii
